src/addons/addonmessage.cpp

```diff
@@ -74,35 +74,97 @@ AddonMessage::AddonMessage(QObject* parent, const QString& manifestFileName,
 
 AddonMessage::~AddonMessage() { MVPN_COUNT_DTOR(AddonMessage); }
 
-void AddonMessage::dismiss() {
-  m_dismissed = true;
-  disable();
-
+// static
+AddonMessage::State AddonMessage::loadMessageState(const QString& id) {
   SettingsHolder* settingsHolder = SettingsHolder::instance();
-  Q_ASSERT(settingsHolder);
 
-  QStringList dismissedAddonMessages = settingsHolder->dismissedAddonMessages();
-  dismissedAddonMessages.append(id());
-  settingsHolder->setDismissedAddonMessages(dismissedAddonMessages);
-}
+  QStringList notifiedAddonMessages = settingsHolder->notifiedAddonMessages();
+  if (notifiedAddonMessages.contains(id)) {
+    return State::Notified;
+  }
 
-void AddonMessage::maskAsRead() {
-  if (m_isRead) {
-    return;
+  QStringList readAddonMessages = settingsHolder->readAddonMessages();
+  if (readAddonMessages.contains(id)) {
+    return State::Read;
   }
 
-  m_isRead = true;
+  QStringList dismissedAddonMessages = settingsHolder->dismissedAddonMessages();
+  if (dismissedAddonMessages.contains(id)) {
+    return State::Dismissed;
+  }
+
+  return State::Received;
+}
+
+void AddonMessage::updateMessageState(State newState) {
+  if (m_state == newState) return;
 
   SettingsHolder* settingsHolder = SettingsHolder::instance();
-  Q_ASSERT(settingsHolder);
 
-  QStringList readAddonMessages = settingsHolder->readAddonMessages();
-  readAddonMessages.append(id());
-  settingsHolder->setReadAddonMessages(readAddonMessages);
+  switch (m_state) {
+    case State::Received: {
+      // Do nothing. This state is not saved in settings,
+      // but it's also not an error.
+      break;
+    }
+    case State::Notified: {
+      QStringList notifiedAddonMessages =
+          settingsHolder->notifiedAddonMessages();
+      notifiedAddonMessages.removeAll(id());
+      settingsHolder->setNotifiedAddonMessages(notifiedAddonMessages);
```

Comment:
if we do a setting group per addon, this code can be simplified a lot. Thoughts?

Reply:
Yeap, agreed. I did not know about settings groups when implementing. Will make the change.